### PR TITLE
Kp/new func multilabel

### PR DESCRIPTION
**Pull request by kprokofi**
- [x] new models (Resnet, TresNet, Wide Resnet, Resnext)

- [x] new loss for multilabel based on the AMSoftmax

- [x] DML for multilabel classification

- [x] support of the EMA for multilabel classification

- [x] new augmentations (cutout, randaugment)
- [x]  Create new engine for multilabel classification

### CODE REVIEW

**Comment by kprokofi** [comment]
@sovrasov , ready to review

File: torchreid/data/datasets/image/classification.py [hunk 1]
```diff
@@ -243,6 +243,8 @@ def load_annotation(annot_path, data_dir, dataset_id=0):
                 rel_image_path, img_labels = img_info
                 full_image_path = osp.join(data_dir, rel_image_path)
                 labels_idx = [class_to_idx[lbl] for lbl in img_labels if lbl in class_to_idx]
+                if not full_image_path or not labels_idx:
```
Comment (sovrasov) [inline torchreid/data/datasets/image/classification.py]: Empty label set should be considered as a negative sample

File: requirements.txt [hunk 2]
```diff
@@ -26,3 +26,5 @@ torch==1.8.1
 optuna
 timm
 addict
+randaugment
+inplace-abn
```
Comment (sovrasov) [inline requirements.txt]: Since we are going to integrate this project into a bigger one, it's better to avoid any buildable dependencies if we can.
I suggest creating optional-requirements.txt and making TResNet available only if this package is installed.

File: torchreid/models/tresnet.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+class TResnetTimm(ModelInterface):
+    def __init__(self,
+                pretrained=False,
+                dropout_cls = None,
+                lr_finder=None,
+                num_classes=1000,
+                **kwargs):
+        super().__init__(**kwargs)
+        dropout_rate = dropout_cls.p
```
Comment (sovrasov) [inline torchreid/models/tresnet.py:14]: Why do dropout implementations for ResNet and TResNet are different?

Comment (kprokofi) [reply torchreid/models/tresnet.py:14]: This is due to timm repo interface. In the case of Resnet we can use our Dropout implementation, but in case of TResnet we cannot as it merged into the head

Comment (kprokofi) [reply torchreid/models/tresnet.py:14]: And I decided to use our Dropout whenever it is possible.

File: requirements.txt (hunk 2 above)
Comment (kprokofi) [reply requirements.txt]: ok

File: torchreid/data/datasets/image/classification.py (hunk 1 above)
Comment (kprokofi) [reply torchreid/data/datasets/image/classification.py]: Yep, I will do that

File: torchreid/engine/image/multilabel.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+        total_loss /= float(num_trg_losses)
+        avg_acc /= float(num_trg_losses)
+
+        return total_loss, loss_summary, avg_acc, out_logits
+
+    def kl_div_binary(self, x, y):
+        ''' compute KL divergence between two tensors represented
+        independent binary distributions'''
+        y_log = torch.log(y)
```
Comment (sovrasov) [inline torchreid/engine/image/multilabel.py]: If you pass targets in log space, you have to use the corresponding flag.

File: torchreid/engine/image/multilabel.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        # mean(1) - mean over binary distributions first
+        # mean(0) - mean over batch then
+        return F.kl_div(y_log, x, reduction='none').mean(1).mean(0)
```
Comment (sovrasov) [inline torchreid/engine/image/multilabel.py]: Seems like this formula is incorrect: kl_div expects that you have both p and 1-p in the input tensor.
Also log target comes as the second arg.

File: torchreid/engine/image/am_softmax.py [hunk 6]
```diff
@@ -247,10 +239,16 @@ def forward_backward(self, data):
                         continue
 
                     with torch.no_grad():
-                        trg_probs = torch.softmax(torch.stack(out_logits[trg_id]), dim=2).mean(dim=0)
+                        if self.loss_name in ['asl', 'am_asl']:
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py]: Is this branch is supposed to be moved to the corresponding engine?

File: torchreid/engine/image/multilabel.py (hunk 4 above)
Comment (kprokofi) [reply torchreid/engine/image/multilabel.py]: If I understand correctly the documentation: https://pytorch.org/docs/1.9.0/generated/torch.nn.KLDivLoss.html#torch.nn.KLDivLoss
KLdivLoss(x,y) where x must be log probs and y must be just probabilities. Maybe I messed up the order. But I think it doesn't matter as we compute KL(x | y) and KL(y | x). I will change the x and y position.

File: torchreid/engine/image/multilabel.py (hunk 5 above)
Comment (kprokofi) [reply torchreid/engine/image/multilabel.py]: We have two distributions P and Q and compute kldiv function (p * (log(p) -log(q)). When we use flag reduction = None, the function applies kldiv to each pair of elements in tensors (what we exactly want, I guess).  Then I mean over classes (binary distributions) for each sample and then mean it over batch. Am I incorrect?

File: torchreid/engine/image/am_softmax.py (hunk 6 above)
Comment (kprokofi) [reply torchreid/engine/image/am_softmax.py]: I remember deleting it... Okay, I will check one more time if the  ASL loss is still present in the AM-Softmax engine

File: torchreid/engine/image/multilabel.py (hunk 5 above)
Comment (sovrasov) [reply torchreid/engine/image/multilabel.py]: P is not a distribution itself, since the considered random variable has two values.

Comment (kprokofi) [reply torchreid/engine/image/multilabel.py]: I got it. I need to compute 1-p as well to form a binary distribution that summed up to 1. Thanks

**Comment by sovrasov** [comment]
@kprokofi please also fix training with AMSoftmax

File: torchreid/data/datasets/image/classification.py [hunk 7]
```diff
@@ -239,10 +239,15 @@ def load_annotation(annot_path, data_dir, dataset_id=0):
...
             images_info = annotation['images']
+            img_wo_objects = 0
             for img_info in images_info:
                 rel_image_path, img_labels = img_info
                 full_image_path = osp.join(data_dir, rel_image_path)
                 labels_idx = [class_to_idx[lbl] for lbl in img_labels if lbl in class_to_idx]
-                assert full_image_path and labels_idx
+                assert full_image_path
+                if not labels_idx:
+                    img_wo_objects += 1
```
Comment (sovrasov) [inline torchreid/data/datasets/image/classification.py:249]: Did you check that an empty tuple will be processed correctly in `getitem`?

File: torchreid/engine/image/am_softmax.py [hunk 8]
```diff
@@ -56,12 +55,14 @@ def __init__(self, datamanager, models, optimizers, reg_cfg, metric_cfg, schedul
...
+                                                   lr_finder=lr_finder,
                                                    use_ema_decay=use_ema_decay,
                                                    ema_decay=ema_decay)
 
-        assert softmax_type in ['softmax', 'am', 'asl']
+        assert loss_name in ['softmax', 'am']
+        self.loss_name = loss_name
         assert s > 0.0
-        if softmax_type == 'am':
+        if loss_name == 'am':
```
Comment (sovrasov) [inline torchreid/engine/image/am_softmax.py]: Here is supposed to be `am_softmax`

File: torchreid/engine/image/multilabel.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+                self.main_losses.append(AsymmetricLoss(
+                    gamma_neg=asl_gamma_neg,
+                    gamma_pos=asl_gamma_pos,
+                    probability_margin=asl_p_m,
+                ))
+            elif loss_name == 'bce':
+                self.main_losses.append(AsymmetricLoss(
+                    gamma_neg=0,
+                    gamma_pos=0,
+                    probability_margin=asl_p_m,
```
Comment (sovrasov) [inline torchreid/engine/image/multilabel.py]: For pure BCE probability_margin is 0

File: torchreid/engine/image/multilabel.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+                    gamma_pos=asl_gamma_pos,
+                ))
+
+        num_classes = self.datamanager.num_train_pids
+        if not isinstance(num_classes, (list, tuple)):
+            num_classes = [num_classes]
+        self.num_classes = num_classes
+        self.num_targets = len(self.num_classes)
+        self.enable_sam = enable_sam
+
```
Comment (sovrasov) [inline torchreid/engine/image/multilabel.py:79]: Del extra line

File: torchreid/data/datasets/image/classification.py (hunk 7 above)
Comment (kprokofi) [reply torchreid/data/datasets/image/classification.py:249]: yes, everything works as expected. 
image
ones are simply not written here.
